File: data_structure/app/src/main/java/com/etiantian/datastructure/MainActivity.java
package com.etiantian.datastructure;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        int f1 = f(1);
        int f2 = f(2);
        int f3 = f(3);
        int f10 = f(10);
        Log.d("1:", "f1:" + f1);
        Log.d("1:", "f2:" + f2);
        Log.d("1:", "f3:" + f3);
        Log.d("1:", "f10:" + f10);
    }

    ///
    /// @description 递归思想演练
    /// @param 
    /// @return 
    /// @author waitwalker
    /// @time 11/20/20 8:36 AM
    ///
    private int f(int x) {
        if (x == 0) {
            return 0;
        }
        return 2 * f(x - 1) + x * x;
    }
}

File: AppA/app/src/main/java/com/sistalk/appa/MainActivity.kt
package com.sistalk.appa

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.sistalk.appa.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private lateinit var mainBinding: ActivityMainBinding

    @SuppressLint("QueryPermissionsNeeded")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        mainBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(mainBinding.root)
        //setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        // 隐式intent跳转
        mainBinding.jumpButton.setOnClickListener {
            // 隐式跳转
            val mIntent = Intent(Intent.ACTION_VIEW, Uri.parse("https://waitwalker.cn"))
            if (mIntent.resolveActivity(packageManager) != null) {
                startActivity(mIntent)
            } else {
                Toast.makeText(this,"App 没有安装，不能处理跳转", Toast.LENGTH_LONG).show()
            }
        }

        // 显示intent跳转
        mainBinding.button2.setOnClickListener {
            val mIntent = Intent()
            mIntent.component = ComponentName("com.sistalk.appb","com.sistalk.appb.SecondActivity")
            if (mIntent.resolveActivity(packageManager) != null) {
                startActivity(mIntent);
            } else {
                Toast.makeText(this, "App B is not installed.", Toast.LENGTH_LONG).show();
            }
        }

        // 通过package manager 涉及到隐私权限
        mainBinding.buttonPackage.setOnClickListener {
            val packageName = "com.sistalk.appc"
            if (isAppInstalled(packageName)) {
                val mIntent = packageManager.getLaunchIntentForPackage(packageName)
                if (mIntent != null) {
                    mIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(mIntent)
                }
            } else {
                Toast.makeText(this,"$packageName 没有安装",Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun isAppInstalled(packageName:String): Boolean{
        var installed = false
        try {
            val packageInfo = packageManager.getPackageInfo(packageName,PackageManager.GET_ACTIVITIES)
            if (packageInfo.packageName == packageName) {
                installed = true
            }
        } catch (e:Exception) {
            installed = false
        } finally {
            Log.d("MainActivity","应用安装状态:$installed")
        }
        return installed
    }
}